parsing/extensions.ml

```diff
@@ -1,175 +1,462 @@
-        let guard = unwrap_expression ~loc (str_hd.pstr_desc) in
-        str_tl, Some guard
-      | _ -> raise(Error(loc, Illegal_comprehension_extension_construct))
-    in
-    let clauses =
-      List.map (fun {pstr_desc; pstr_loc=_;}  ->
-          match (unwrap_expression ~loc pstr_desc).pexp_desc with
-          | Pexp_for(p, e2, e3, dir, _) -> From_to (p, e2, e3, dir)
-          | Pexp_let(Nonrecursive,
-          [{
-            pvb_pat=p;
-            pvb_expr=e2;
-            pvb_attributes=_;
-            pvb_loc=_;
-          }], _) ->  In (p, e2)
-          | _ -> raise(Error(loc, Illegal_comprehension_extension_construct))
-        ) str
+          match ext_payload with
+          | PStr [] -> begin
+              match arguments with
+              | [Nolabel, body] -> Some (names, body)
+              | _ -> raise_malformed (Wrong_arguments arguments)
+            end
+          | _ -> raise_malformed (Has_payload ext_payload)
+        end
+      | _ -> None
+    end
+  | _ -> None
+
+module Comprehensions = struct
+  type iterator =
+    | Range of { start     : expression
+               ; stop      : expression
+               ; direction : Asttypes.direction_flag }
+        (** "= START to STOP" (direction = Upto)
+            "= START downto STOP" (direction = Downto) *)
+    | In of expression
+      (** "in EXPR" *)
+
+  type clause_binding =
+    { pattern    : pattern
+    ; iterator   : iterator
+    ; attributes : attribute list }
+    (** PAT (in/= ...) [@...] *)
+
+  type clause =
+    | For of clause_binding list
+        (** "for PAT (in/= ...) and PAT (in/= ...) and ..."; must be nonempty *)
+    | When of expression
+        (** "when EXPR" *)
+
+  type comprehension =
+    { body    : expression
+        (** The body/generator of the comprehension *)
+    ; clauses : clause list
+        (** The clauses of the comprehension; must be nonempty *) }
+
+  type comprehension_expr =
+    | Cexp_list_comprehension  of comprehension
+        (** [BODY ...CLAUSES...] *)
+    | Cexp_array_comprehension of comprehension
+        (** [|BODY ...CLAUSES...|] *)
+
+  let extension_name = Clflags.Extension.to_string Comprehensions
+
+  (* CR aspectorzabusky: new name? *)
+  let comprehension_expr ~loc names =
+    extension_expr ~loc (extension_name :: names)
+
+  let expr_of_iterator ~loc = function
+    | Range { start; stop; direction } ->
+        comprehension_expr
+          ~loc
+          [ "for"
+          ; "range"
+          ; match direction with
+            | Upto   -> "upto"
+            | Downto -> "downto" ]
+          (Ast_helper.Exp.tuple [start; stop])
+    | In seq ->
+        comprehension_expr ~loc ["for"; "in"] seq
+
+  let expr_of_clause_binding ~loc { pattern; iterator; attributes } =
+    Ast_helper.Vb.mk
+      ~loc
+      ~attrs:attributes
+      pattern
+      (expr_of_iterator ~loc iterator)
+
+  let expr_of_clause ~loc = function
+    | For iterators -> fun rest ->
```

Comment:
I think so I could use `function` rather than an explicit match, or maybe to avoid closing over things?  Also possibly just code changes over time?  I don't recall specifically.

Reply:
While `function` is indeed nice, it's probably better not to use it here. Might interfere with e.g. arity analysis, and it makes the code harder to read with the parameters scattered over three different places (to the left of the `=`, the one parameter eaten by `function`, and `rest`).